src/bun.js/bindings/NodeVMScript.cpp

```diff
@@ -0,0 +1,646 @@
+        JSC::SourceCodeKey key(source, {}, JSC::SourceCodeType::ProgramType, lexicallyScopedFeatures, JSC::JSParserScriptMode::Classic, JSC::DerivedContextType::None, JSC::EvalContextType::None, false, {}, std::nullopt);
+        Ref<JSC::CachedBytecode> cachedBytecode = JSC::CachedBytecode::create(std::span(cachedData), nullptr, {});
+        JSC::UnlinkedProgramCodeBlock* unlinkedBlock = JSC::decodeCodeBlock<UnlinkedProgramCodeBlock>(vm, key, WTFMove(cachedBytecode));
+
+        if (!unlinkedBlock) {
+            script->cachedDataRejected(TriState::True);
+        } else {
+            JSC::JSScope* jsScope = globalObject->globalScope();
+            JSC::CodeBlock* codeBlock = nullptr;
+            {
+                // JSC::ProgramCodeBlock::create() requires GC to be deferred.
+                DeferGC deferGC(vm);
+                codeBlock = JSC::ProgramCodeBlock::create(vm, executable, unlinkedBlock, jsScope);
+            }
+            JSC::CompilationResult compilationResult = JIT::compileSync(vm, codeBlock, JITCompilationEffort::JITCompilationCanFail);
+            if (compilationResult != JSC::CompilationResult::CompilationFailed) {
+                executable->installCode(codeBlock);
+                script->cachedDataRejected(TriState::False);
+            } else {
+                script->cachedDataRejected(TriState::True);
+            }
+        }
+    } else if (produceCachedData) {
+        script->cacheBytecode();
+        // TODO(@heimskr): is there ever a case where bytecode production fails?
+        script->cachedDataProduced(true);
+    }
+
+    return JSValue::encode(script);
+}
+
+JSC_DEFINE_HOST_FUNCTION(scriptConstructorCall, (JSGlobalObject * globalObject, CallFrame* callFrame))
+{
+    return constructScript(globalObject, callFrame);
+}
+
+JSC_DEFINE_HOST_FUNCTION(scriptConstructorConstruct, (JSGlobalObject * globalObject, CallFrame* callFrame))
+{
+    return constructScript(globalObject, callFrame, callFrame->newTarget());
+}
+
+JSC::ProgramExecutable* NodeVMScript::createExecutable()
+{
+    VM& vm = JSC::getVM(globalObject());
+    m_cachedExecutable.set(vm, this, JSC::ProgramExecutable::create(globalObject(), m_source));
+    return m_cachedExecutable.get();
+}
+
+void NodeVMScript::cacheBytecode()
+{
+    if (!m_cachedExecutable) {
+        createExecutable();
+    }
+
+    m_cachedBytecode = getBytecode(globalObject(), m_cachedExecutable.get(), m_source);
+    m_cachedDataProduced = m_cachedBytecode != nullptr;
+}
+
+JSC::JSUint8Array* NodeVMScript::getBytecodeBuffer()
+{
+    if (!m_options.produceCachedData) {
+        return nullptr;
+    }
+
+    if (!m_cachedBytecodeBuffer) {
+        if (!m_cachedBytecode) {
+            cacheBytecode();
+        }
+
+        ASSERT(m_cachedBytecode);
+
+        std::span<const uint8_t> bytes = m_cachedBytecode->span();
+        m_cachedBytecodeBuffer.set(vm(), this, WebCore::createBuffer(globalObject(), bytes));
```

Comment:
this needs an exception scope technically
```suggestion
        m_cachedBytecodeBuffer.set(vm(), this, WebCore::createBuffer(globalObject(), bytes));
```